File: src/cpp/server/request_handler/resp_parser.cpp
#include <cstdint>

#include "resp_parser.h"
#include "utils.h"


std::optional<RespValue> RespParser::parse(const std::vector<uint8_t>& buf) {
    return parse(std::string_view(reinterpret_cast<const char*>(buf.data()), buf.size()));
}

std::optional<RespValue> RespParser::parse(std::string_view buf) {
    if (buf.empty()) {
        return std::nullopt;
    }

    switch (buf[0]) {
        case '+': return parse_simple_string(buf);
        case '-': return parse_error(buf);
        case ':': return parse_integer(buf);
        case '$': return parse_bulk_string(buf);
        case '*': return parse_array(buf);
        default:  return RespValue::error("ERR unknown RESP type byte");
    }
}

std::optional<RespValue> RespParser::parse_simple_string(std::string_view buf) {
    if (buf[0] != '+') {
        return RespValue::error("ERR invalid simple string format");
    }

    size_t pos = find_crlf(buf);
    if (pos == std::string_view::npos) {
        return std::nullopt;
    }
    std::string_view value = buf.substr(1, pos - 1);
    return RespValue::simple_string(std::string(value), pos + 2);
}

std::optional<RespValue> RespParser::parse_error(std::string_view buf) {
    if (buf[0] != '-') {
        return RespValue::error("ERR invalid error format");
    }

    size_t pos = find_crlf(buf);
    if (pos == std::string_view::npos) {
        return std::nullopt;
    }
    std::string_view value = buf.substr(1, pos - 1);
    return RespValue::error(std::string(value), pos + 2);
}

std::optional<RespValue> RespParser::parse_integer(std::string_view buf) {
    if (buf[0] != ':') {
        return RespValue::error("ERR invalid integer format");
    }

    if (buf.size() < 3) {
        return std::nullopt;
    }

    size_t pos = find_crlf(buf);
    if (pos == std::string_view::npos) {
        return std::nullopt;
    }

    int value;
    if (!parse_string_as_int(buf.substr(1, pos - 1), value)) {
        return RespValue::error("ERR invalid integer format");
    }

    return RespValue::integer(value, pos + 2);
}

std::optional<RespValue> RespParser::parse_bulk_string(std::string_view buf) {
    if (buf[0] != '$') {
        return RespValue::error("ERR invalid bulk string format");
    }

    size_t pos_1 = find_crlf(buf);
    if (pos_1 == std::string_view::npos) {
        return std::nullopt;
    }

    int length;
    if (!parse_string_as_int(buf.substr(1, pos_1 - 1), length)) {
        return RespValue::error("ERR invalid bulk string format");
    }

    if (length == -1) {
        return RespValue::null_bulk(pos_1 + 2);
    }

    if (length < 0) {
        return RespValue::error("ERR invalid bulk string format");
    }

    size_t data_start = pos_1 + 2;
    size_t total_bytes = data_start + static_cast<size_t>(length) + 2;

    if (buf.size() < total_bytes) {
        return std::nullopt;
    }

    if (buf[data_start + length] != '\r' || buf[data_start + length + 1] != '\n') {
        return RespValue::error("ERR invalid bulk string format");
    }

    std::string_view value = buf.substr(data_start, static_cast<size_t>(length));
    return RespValue::bulk_string(std::string(value), total_bytes);
}

std::optional<RespValue> RespParser::parse_array(std::string_view buf) {
    if (buf[0] != '*') {
        return RespValue::error("ERR invalid array format");
    }

    size_t pos_1 = find_crlf(buf);
    if (pos_1 == std::string_view::npos) {
        return std::nullopt;
    }

    int num_elements;
    if (!parse_string_as_int(buf.substr(1, pos_1 - 1), num_elements)) {
        return RespValue::error("ERR invalid array format");
    }

    if (num_elements == -1) {
        return RespValue::null_array(pos_1 + 2);
    }

    if (num_elements < 0) {
        return RespValue::error("ERR invalid array format");
    }

    size_t curr_pos = pos_1 + 2;
    std::vector<RespValue> elements;
    for (int i = 0; i < num_elements; i++) {
        std::optional<RespValue> element = parse(buf.substr(curr_pos));
        if (!element.has_value()) {
            return std::nullopt;
        }
        RespValue v = element.value();
        if (v.num_resp_bytes == RespValue::kUnsetRespBytes) {
            return RespValue::error("ERR invalid array element");
        }
        elements.push_back(v);
        curr_pos += v.num_resp_bytes;
    }

    return RespValue::array_from_parts(elements, curr_pos);
}

size_t RespParser::find_crlf(std::string_view buf) {
    // returns 0-indexed position of the first "\r\n", or npos if not found.
    size_t pos = buf.find("\r\n");
    if (pos == std::string_view::npos) {
        return std::string_view::npos;
    }
    return pos;
}

File: src/python/resp_parser.py
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Optional


class RespType(Enum):
    SimpleString = "SimpleString"
    Error = "Error"
    Integer = "Integer"
    BulkString = "BulkString"
    Array = "Array"
    Null = "Null"


@dataclass
class RespValue:
    type: RespType
    str_value: str = ""
    int_value: int = 0
    array: list["RespValue"] = field(default_factory=list)
    num_resp_bytes: Optional[int] = None
    is_null: bool = False

    @staticmethod
    def simple_string(s: str, num_resp_bytes: int) -> "RespValue":
        return RespValue(type=RespType.SimpleString, str_value=s, num_resp_bytes=num_resp_bytes)

    @staticmethod
    def error(msg: str, num_resp_bytes: Optional[int] = None) -> "RespValue":
        return RespValue(type=RespType.Error, str_value=msg, num_resp_bytes=num_resp_bytes)

    @staticmethod
    def integer(n: int, num_resp_bytes: int) -> "RespValue":
        return RespValue(type=RespType.Integer, int_value=n, num_resp_bytes=num_resp_bytes)

    @staticmethod
    def bulk_string(s: str, num_resp_bytes: int) -> "RespValue":
        return RespValue(type=RespType.BulkString, str_value=s, num_resp_bytes=num_resp_bytes)

    @staticmethod
    def null_bulk(num_resp_bytes: int) -> "RespValue":
        return RespValue(type=RespType.Null, is_null=True, num_resp_bytes=num_resp_bytes)

    @staticmethod
    def null_array(num_resp_bytes: int) -> "RespValue":
        return RespValue(type=RespType.Null, is_null=True, num_resp_bytes=num_resp_bytes)

    @staticmethod
    def array_from_parts(elements: list["RespValue"], num_resp_bytes: int) -> "RespValue":
        return RespValue(type=RespType.Array, array=elements, num_resp_bytes=num_resp_bytes)



def _find_crlf(buf: str) -> int:
    """Returns the index of the first '\\r\\n', or -1 if not found."""
    pos = buf.find("\r\n")
    return pos


def parse(buf: str | bytes) -> Optional[RespValue]:
    if isinstance(buf, (bytes, bytearray)):
        buf = buf.decode()

    if not buf:
        return None

    match buf[0]:
        case '+': return _parse_simple_string(buf)
        case '-': return _parse_error(buf)
        case ':': return _parse_integer(buf)
        case '$': return _parse_bulk_string(buf)
        case '*': return _parse_array(buf)
        case _:   return RespValue.error("ERR unknown RESP type byte")


def _parse_simple_string(buf: str) -> Optional[RespValue]:
    if buf[0] != '+':
        return RespValue.error("ERR invalid simple string format")

    pos = _find_crlf(buf)
    if pos == -1:
        return None

    value = buf[1:pos]
    return RespValue.simple_string(value, pos + 2)


def _parse_error(buf: str) -> Optional[RespValue]:
    if buf[0] != '-':
        return RespValue.error("ERR invalid error format")

    pos = _find_crlf(buf)
    if pos == -1:
        return None

    value = buf[1:pos]
    return RespValue.error(value, pos + 2)


def _parse_integer(buf: str) -> Optional[RespValue]:
    if buf[0] != ':':
        return RespValue.error("ERR invalid integer format")

    if len(buf) < 3:
        return None

    pos = _find_crlf(buf)
    if pos == -1:
        return None

    try:
        value = int(buf[1:pos])
    except ValueError:
        return RespValue.error("ERR invalid integer format")

    return RespValue.integer(value, pos + 2)


def _parse_bulk_string(buf: str) -> Optional[RespValue]:
    if buf[0] != '$':
        return RespValue.error("ERR invalid bulk string format")

    pos_1 = _find_crlf(buf)
    if pos_1 == -1:
        return None

    try:
        length = int(buf[1:pos_1])
    except ValueError:
        return RespValue.error("ERR invalid bulk string format")

    if length == -1:
        return RespValue.null_bulk(pos_1 + 2)

    if length < 0:
        return RespValue.error("ERR invalid bulk string format")

    data_start = pos_1 + 2
    total_bytes = data_start + length + 2

    if len(buf) < total_bytes:
        return None

    if buf[data_start + length] != '\r' or buf[data_start + length + 1] != '\n':
        return RespValue.error("ERR invalid bulk string format")

    value = buf[data_start:data_start + length]
    return RespValue.bulk_string(value, total_bytes)


def _parse_array(buf: str) -> Optional[RespValue]:
    if buf[0] != '*':
        return RespValue.error("ERR invalid array format")

    pos_1 = _find_crlf(buf)
    if pos_1 == -1:
        return None

    try:
        num_elements = int(buf[1:pos_1])
    except ValueError:
        return RespValue.error("ERR invalid array format")

    if num_elements == -1:
        return RespValue.null_array(pos_1 + 2)

    if num_elements < 0:
        return RespValue.error("ERR invalid array format")

    curr_pos = pos_1 + 2
    elements = []
    for _ in range(num_elements):
        element = parse(buf[curr_pos:])
        if element is None:
            return None
        if element.num_resp_bytes is None:
            return RespValue.error("ERR invalid array element")
        elements.append(element)
        curr_pos += element.num_resp_bytes

    return RespValue.array_from_parts(elements, curr_pos)
